storage/bootstrap/types.go

```diff
@@ -26,22 +26,35 @@ import (
 	xtime "github.com/m3db/m3x/time"
 )
 
+// ProcessProvider constructs a bootstrap process that can execute a
+// bootstrap run.
+type ProcessProvider interface {
+	// SetBootstrapper sets the bootstrapper provider to use when running the
+	// process.
+	SetBootstrapperProvider(bootstrapper BootstrapperProvider)
+
+	// Bootstrapper returns the current bootstrappe provider to use when
+	// running the process.
+	BootstrapperProvider() BootstrapperProvider
+
+	// Provide constructs a bootstrap process.
+	Provide() Process
+}
+
 // Process represents the bootstrap process. Note that a bootstrap process can and will
 // be reused so it is important to not rely on state stored in the bootstrap itself
 // with the mindset that it will always be set to default values from the constructor.
 type Process interface {
-	// SetBootstrapper sets the bootstrapper to use when running the process.
-	SetBootstrapper(bootstrapper Bootstrapper)
-
-	// Bootstrapper returns the current bootstrapper to use when running the process.
-	Bootstrapper() Bootstrapper
-
 	// Run runs the bootstrap process, returning the bootstrap result and any error encountered.
 	Run(
 		ns namespace.Metadata,
 		shards []uint32,
 		targetRanges []TargetRange,
-	) (result.BootstrapResult, error)
+	) (
```

Comment:
think we need to take separate targetRanges for data v index. The time calculations depend on the block size so we could end up different values - 
![screen shot 2018-05-04 at 01 30 55](https://user-images.githubusercontent.com/199982/39614112-a3558b7c-4f5c-11e8-95d9-bf3b28bea46c.png)

Reply:
I was thinking we would be looking at the namespace index options block size and be using the cache from the previous target range to stitch the data together.  But true it'd be more clean to just issue two different target ranges.  

I might even move this logic to the bootstrap process itself, it seems a little odd for it to be here.